File: backend/src/main/java/it/chalmers/gamma/app/oauth2/GammaRegisteredClientRepository.java
package it.chalmers.gamma.app.oauth2;

import it.chalmers.gamma.app.client.domain.*;
import org.springframework.security.oauth2.core.AuthorizationGrantType;
import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
import org.springframework.security.oauth2.server.authorization.settings.ClientSettings;
import org.springframework.stereotype.Component;

@Component
public class GammaRegisteredClientRepository implements RegisteredClientRepository {

    private final ClientRepository clientRepository;

    public GammaRegisteredClientRepository(ClientRepository clientRepository) {
        this.clientRepository = clientRepository;
    }

    @Override
    public void save(RegisteredClient registeredClient) {
        throw new UnsupportedOperationException("Use ClientFacade instead.");
    }

    @Override
    public RegisteredClient findById(String id) {
        Client client = this.clientRepository.get(ClientUid.valueOf(id))
                .orElseThrow(NullPointerException::new);

        return toRegisteredClient(client);
    }

    @Override
    public RegisteredClient findByClientId(String clientId) {
        Client client = this.clientRepository.get(new ClientId(clientId))
                .orElseThrow(NullPointerException::new);

        return toRegisteredClient(client);
    }

    private RegisteredClient toRegisteredClient(Client client) {
        RegisteredClient.Builder builder = RegisteredClient
                .withId(client.clientUid().getValue())
                .clientId(client.clientId().value())
                .clientSecret(client.clientSecret().value())
                .redirectUri(client.redirectUrl().value())
                .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
                .clientName(client.prettyName().value())
                .clientSettings(
                        ClientSettings
                                .builder()
                                .requireAuthorizationConsent(true)
                                .build()
                );

        builder.scope("openid");

        for (Scope scope : client.scopes()) {
            builder.scope(scope.name().toLowerCase());
        }

        return builder.build();
    }

}

File: contexts/oauth-authorization-server/src/main/kotlin/it/chalmers/gamma/oauth/server/GammaRegisteredClientRepository.kt
package it.chalmers.gamma.oauth.server

import it.chalmers.gamma.oauth.ClientId
import it.chalmers.gamma.oauth.ClientOwner
import it.chalmers.gamma.oauth.ClientUid
import it.chalmers.gamma.oauth.OAuthProtocolClients
import it.chalmers.gamma.oauth.OAuthServerClient
import org.springframework.security.oauth2.core.AuthorizationGrantType
import org.springframework.security.oauth2.core.ClientAuthenticationMethod
import org.springframework.security.oauth2.server.authorization.client.RegisteredClient
import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository
import org.springframework.security.oauth2.server.authorization.settings.ClientSettings

internal class GammaRegisteredClientRepository(
    private val clients: OAuthProtocolClients,
) : RegisteredClientRepository {
    override fun save(registeredClient: RegisteredClient): Unit =
        throw UnsupportedOperationException("Clients are managed by the OAuth bounded context")

    override fun findById(id: String): RegisteredClient? =
        runCatching { ClientUid.parse(id) }.getOrNull()?.let { clients.serverClient(it) }?.toSpringClient()

    override fun findByClientId(clientId: String): RegisteredClient? =
        runCatching { ClientId(clientId) }.getOrNull()?.let { clients.serverClient(it) }?.toSpringClient()

    private fun OAuthServerClient.toSpringClient(): RegisteredClient {
        val source = client
        val builder =
            RegisteredClient
                .withId(source.uid.value.toString())
                .clientId(source.clientId.value)
                .clientSecret(encodedSecret)
                .clientName(source.name.value)
                .redirectUri(source.redirectUri.value)
                .clientAuthenticationMethod(ClientAuthenticationMethod.CLIENT_SECRET_BASIC)
                .clientAuthenticationMethod(ClientAuthenticationMethod.CLIENT_SECRET_POST)
                .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
                .clientSettings(
                    ClientSettings
                        .builder()
                        .requireAuthorizationConsent(true)
                        .requireProofKey(false)
                        .setting("gamma.is-official", source.owner is ClientOwner.Official)
                        .build(),
                )
        source.scopes.forEach { builder.scope(it.wireValue) }
        return builder.build()
    }
}
